packages/indy-vdr/src/pool/IndyVdrPoolService.ts

```diff
@@ -0,0 +1,230 @@
+import type { AgentContext } from '../../../core/src/agent'
+import { GetNymRequest, indyVdr,} from 'indy-vdr-test-shared'
+
+import {
+  AgentDependencies,
+  Logger,
+  InjectionSymbols,
+  injectable,
+  inject,
+  LedgerError,
+  LedgerNotConfiguredError,
+  LedgerNotFoundError,
+} from '@aries-framework/core'
+
+import { IndyVdrError, GetNymResponse } from 'indy-vdr-test-shared'
+import { CacheRepository, PersistedLruCache } from '../../../core/src/cache'
+import { IndySdkError } from '../../../core/src/error'
+import { isSelfCertifiedDid } from '../../../core/src/utils/did'
+import { isIndyError } from '../../../core/src/utils/indyError'
+import { allSettled, onlyFulfilled, onlyRejected } from '../../../core/src/utils/promises'
+import { assertIndyWallet } from '../../../core/src/wallet/util/assertIndyWallet'
+import { IndyVdrPool } from './IndyVdrPool'
+
+export const DID_POOL_CACHE_ID = 'DID_POOL_CACHE'
+export const DID_POOL_CACHE_LIMIT = 500
+export interface CachedDidResponse {
+  nymResponse: {
+    did: string
+    verkey: string
+  }
+  poolId?: string
+}
+@injectable()
+export class IndyVdrPoolService {
+  public pools: IndyVdrPool[] = []
+  private logger: Logger
+  private indyVdr!: typeof indyVdr
+  private agentDependencies: AgentDependencies
+  private didCache: PersistedLruCache<CachedDidResponse>
+
+  public constructor(
+    cacheRepository: CacheRepository,
+    @inject(InjectionSymbols.AgentDependencies) agentDependencies: AgentDependencies,
+    @inject(InjectionSymbols.Logger) logger: Logger
+  ) {
+    this.logger = logger
+    this.agentDependencies = agentDependencies
+
+    this.didCache = new PersistedLruCache(DID_POOL_CACHE_ID, DID_POOL_CACHE_LIMIT, cacheRepository)
+  }
+
+  /**
+   * Create connections to all ledger pools
+   */
+  public async connectToPools() {
+    const handleArray: number[] = []
+    // Sequentially connect to pools so we don't use up too many resources connecting in parallel
+    for (const pool of this.pools) {
+      this.logger.debug(`Connecting to pool: ${pool.id}`)
+      const poolHandle = await pool.connect()
+      this.logger.debug(`Finished connection to pool: ${pool.id}`)
+      handleArray.push(poolHandle)
+    }
+    return handleArray
+  }
+
+  /**
+   * Get the most appropriate pool for the given did. The algorithm is based on the approach as described in this document:
+   * https://docs.google.com/document/d/109C_eMsuZnTnYe2OAd02jAts1vC4axwEKIq7_4dnNVA/edit
+   */
+  public async getPoolForDid(agentContext: AgentContext, did: string): Promise<{ pool: IndyVdrPool }> {
+    // Check if the did starts with did:indy
+    if (did.startsWith('did:indy')) {
+      const nameSpace = did.split(':')[2]
```

Comment:
The namespace can contain multiple colons. 

E.g. the following dids are all correct:

- `did:indy:sovrin:builder:xxx` -> namespace = `sovrin:builder`
- `did:indy:bcovrin:xxx` -> namespace = `bcovrin`.

We should proabbly use a regex to match the did. If you store the followign const in a separate file:

```
const DID_INDY_REGEX = /^did:indy:((?:[a-z][_a-z0-9-]*)(?::[a-z][_a-z0-9-]*)):([1-9A-HJ-NP-Za-km-z]{21,22})$/
```

Then you can use it like this:

```ts
const match = did.match(DID_INDY_REGEX)

// If we have a match it means the did starts with did:indy and is a qualified indy did
if (match) {
  const [, namespace] = match
} else {
  // legacy way
}
```

Reply:
We can then also just use `this.getPoolForNamespace` and reuse that logic